### DISCUSSION TITLE
Online Mining Questions

### DISCUSSION THREAD

**Discussion by tasnimtazzy** [talmolab/sleap#1385] [Help!] [answered]
Hello! 

I am starting to train models of octopi by treating each arm as a separate entity and came across the possibility of using Online Hard Keypoint Mining to train models. I found the following code through the SLEAP website: 

class HardKeypointMiningConfig:
    """Configuration for online hard keypoint mining.

    Attributes:
        online_mining: If True, online hard keypoint mining (OHKM) will be enabled. When
            this is enabled, the loss is computed per keypoint (or edge for PAFs) and
            sorted from lowest (easy) to highest (hard). The hard keypoint loss will be
            scaled to have a higher weight in the total loss, encouraging the training
            to focus on tricky body parts that are more difficult to learn.
            If False, no mining will be performed and all keypoints will be weighted
            equally in the loss.
        hard_to_easy_ratio: The minimum ratio of the individual keypoint loss with
            respect to the lowest keypoint loss in order to be considered as "hard".
            This helps to switch focus on across groups of keypoints during training.
        min_hard_keypoints: The minimum number of keypoints that will be considered as
            "hard", even if they are not below the `hard_to_easy_ratio`.
        max_hard_keypoints: The maximum number of hard keypoints to apply scaling to.
            This can help when there are few very easy keypoints which may skew the
            ratio and result in loss scaling being applied to most keypoints, which can
            reduce the impact of hard mining altogether.
        loss_scale: Factor to scale the hard keypoint losses by.
    """

    online_mining: bool = False
    hard_to_easy_ratio: float = 2.0
    min_hard_keypoints: int = 2
    max_hard_keypoints: Optional[int] = None
    loss_scale: float = 5.0

This might be a stupid question but where do I use this piece of code? Do I paste it into the terminal while models are being trained? How can I utilize this code? Any help would be greatly appreciated, thank you!

**Comment by davidasamy** [comment, marked as answer]
This is an option to enable when you are defining your [sleap training profile](https://sleap.ai/develop/guides/custom-training.html). simply hit run and start training through the GUI, or you can export this profile and then use it to run training through the command line.

1. Run sleap with `sleap-label` 
2. Create skeleton & label frames
3. Predict -> Run Training -> Model Configuration -> Enable online mining

**Reply by tasnimtazzy** [reply]
Thank you so much for your help!
